crates/risc0-boundless/src/host.rs

```diff
@@ -0,0 +1,539 @@
+                    // Save request for retries
+                    last_request = Some(req.clone());
+                    req
+                };
+                match request {
+                    BoundlessRequest::UploadImg { elf, notify } => {
+                        let res = client.upload_image(&elf);
+                        let res = unwrap_boundless_response!(res, 'queue);
+                        let _ = notify.send(res);
+                    }
+                    BoundlessRequest::UploadInput { input, notify } => {
+                        debug!("Boundless: upload_input");
+                        let res = client.upload_input(&input);
+                        let res = unwrap_boundless_response!(res, 'queue);
+                        let _ = notify.send(res);
+                    }
+                    BoundlessRequest::SubmitReq {
+                        proving_request,
+                        notify,
+                    } => {
+                        debug!("Boundless: submit_req");
+                        let res = client.submit_request(&proving_request);
+                        let res = unwrap_boundless_response!(res, 'queue);
+                        let _ = notify.send(res);
+                    }
+                    BoundlessRequest::WaitForReqFulfillment {
+                        request_id,
+                        check_interval,
+                        timeout,
+                        notify,
+                    } => {
+                        debug!("Boundless: wait_for_req_fulfillment");
+                        let res = client
+                            .wait_for_request_fulfillment(request_id, check_interval, timeout)
+                            .await;
+                        // There is no need to retry this request as this already has a retry mechanism inside
+                        let _ = notify.send(res);
+                    }
+                    BoundlessRequest::Slash { request_id, notify } => {
+                        debug!("Boundless: slash");
+                        let res = client
+                            .proof_market
+                            .slash(request_id)
+                            .await
+                            .map_err(ClientError::MarketError);
+                        let _ = notify.send(res);
+                    }
+                };
+                // We arrive here only on a successful response
+                last_request = None;
+            }
+        });
+        let _join_handle = Arc::new(join_handle);
+
+        Self {
+            queue,
+            _join_handle,
+        }
+    }
+
+    #[instrument(level = "trace", skip(self), ret)]
+    fn upload_img(&self, elf: Vec<u8>) -> String {
+        let (notify, mut rx) = mpsc::unbounded_channel();
+        self.queue
+            .send(BoundlessRequest::UploadImg { elf, notify })
+            .unwrap();
+        tokio::task::block_in_place(|| rx.blocking_recv().unwrap())
```

Comment:
Why use `block_in_place` when you can make this method async + you're calling this method from an async block?